libraries/chain/finality/qc.cpp

```diff
@@ -309,10 +375,54 @@ std::optional<qc_t> aggregating_qc_t::get_best_qc(block_num_type block_num) cons
    return std::optional<qc_t>{qc_t{block_num, std::move(*active_best_qc), {}}};
 }
 
+// A dual finalizer votes on both active and pending finalizer policies
+void aggregating_qc_t::verify_dual_finalizers_votes(const qc_t& qc) const {
+   auto build_vote_index = [&](const std::vector<finalizer_authority> finalizers, std::set<bls_public_key>& keys, std::map<bls_public_key, uint32_t>& index) {
+      uint32_t i = 0;
+      for (const auto& f: finalizers) {
+         index[f.public_key] = i;
+         keys.insert(f.public_key);
+         ++i;
+      }
+   };
+
+   std::set<bls_public_key> active_finalizer_keys; // all active finalizer public keys
+   std::map<bls_public_key, uint32_t> active_finalizer_vote_index; // from public key to vote bit index
+   build_vote_index(active_finalizer_policy->finalizers,
+                    active_finalizer_keys,
+                    active_finalizer_vote_index);
+
+   std::set<bls_public_key> pending_finalizer_keys; // all pending finalizer public keys
+   std::map<bls_public_key, uint32_t> pending_finalizer_vote_index; // from public key to vote bit index
+   build_vote_index(pending_finalizer_policy->finalizers,
+                    pending_finalizer_keys,
+                    pending_finalizer_vote_index);
+
+   // Find dual finalizers (which vote on both active and pending policies
+   std::set<bls_public_key> dual_finalizer_keys;
+   std::set_intersection(active_finalizer_keys.begin(), active_finalizer_keys.end(),
+                         pending_finalizer_keys.begin(), pending_finalizer_keys.end(),
+                         std::inserter(dual_finalizer_keys, dual_finalizer_keys.begin()));
+
+   // Verify each dual finalizer votes in the same way
+   for (const auto& k: dual_finalizer_keys) {
+      auto active_vote_index  = active_finalizer_vote_index[k];
+      auto pending_vote_index = pending_finalizer_vote_index[k];
+      EOS_ASSERT(qc.vote_same_at(active_vote_index, pending_vote_index),
+                 invalid_qc_claim,
+                 "qc ${bn} contains contains a dual finalizer ${k} which does not vote the same on active and pending policies", ("bn", qc.block_num)("k", k));
+   }
+}
+
 void aggregating_qc_t::verify_qc(const qc_t& qc, const digest_type& strong_digest, const weak_digest_t& weak_digest) const {
+   qc.active_policy_sig.verify_vote_format(active_finalizer_policy);
+
    if (qc.pending_policy_sig) {
       EOS_ASSERT(pending_finalizer_policy, invalid_qc_claim,
                  "qc ${bn} contains pending policy signature for nonexistent pending finalizer policy", ("bn", qc.block_num));
+
+      qc.pending_policy_sig->verify_vote_format(active_finalizer_policy);
```

Comment:
Should be `pending_finalizer_policy`

Reply:
Thanks! (cut and paste...). Fixed by https://github.com/AntelopeIO/spring/pull/475/commits/2841dd00717e70d2ae545cb8cb73c7dee7c356e0